### PR TITLE
[SYCL-MLIR] Raise host construction of sycl::id and sycl::range

### PR DESCRIPTION
**Pull request by sommerlukas**
Add patterns to raise the construction of `sycl::id` and `sycl::range` on the host side. The patterns differ from other patterns, as the constructors are usually inlined, so the pattern needs to match the initialization of the member fields directly.

### CODE REVIEW

**Review by jopperm — APPROVED** [review]
> as the constructors are usually inlined

What would happen if they haven't been inlined?

**Comment by sommerlukas** [comment]
> > as the constructors are usually inlined
> 
> What would happen if they haven't been inlined?

The pattern wouldn't match and we would not perform the raising. As we do not emit the code back out and the raising on the host side is just to facilitate analysis, it should not have any impact on the correctness of the compilation.

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp [hunk 1]
```diff
@@ -291,6 +291,202 @@ struct AccessorInvokeConstructorPattern
...
+template <typename TypeTag>
+class RaiseArrayConstructorBasePattern
+    : public OpRewritePattern<LLVM::StoreOp> {
+public:
+  using OpRewritePattern<LLVM::StoreOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(LLVM::StoreOp op,
+                                PatternRewriter &rewriter) const final {
+    // The 'this*' is the address to which the element stores are performed.
+    auto alloc = dyn_cast_or_null<LLVM::AllocaOp>(op.getAddr().getDefiningOp());
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:303]:
```suggestion
    auto alloc = op.getAddr().getDefiningOp<LLVM::AllocaOp>();
```

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp [hunk 2]
```diff
@@ -291,6 +291,202 @@ struct AccessorInvokeConstructorPattern
...
+
+    if (!alloc)
+      return failure();
+
+    assert(alloc.getElemType().has_value() &&
+           "Expecting element type attribute for opaque alloca");
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:309]: If the device side is using typed pointers, we might hit this assert. Return `failure()` insted.

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp [hunk 3]
```diff
@@ -291,6 +291,202 @@ struct AccessorInvokeConstructorPattern
...
+        !TypeTag::getTypeName().match(structAllocTy.getName()))
+      return failure();
+
+    auto arrayTyOrNone = getNumAndTypeOfComponents(structAllocTy);
+    if (!arrayTyOrNone)
+      // Failed to identify the number of dimensions/components
+      return failure();
+
+    auto numComponents = (*arrayTyOrNone).getNumElements();
+    auto componentTy = (*arrayTyOrNone).getElementType();
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:325]:
```suggestion
    auto numComponents = arrayTyOrNone->getNumElements();
    auto componentTy = arrayTyOrNone->getElementType();
```
NIT

File: polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp [hunk 4]
```diff
@@ -291,6 +291,202 @@ struct AccessorInvokeConstructorPattern
...
+  Operation *
+  findLastComponentInBlock(mlir::Block *block,
+                           llvm::ArrayRef<Component> components) const {
+    llvm::SmallPtrSet<Operation *, 3> stores;
+    for (const auto &c : components) {
+      stores.insert(c.store);
+    }
+    auto lastComponent =
+        std::find_if(block->rbegin(), block->rend(),
+                     [&](Operation &op) { return stores.contains(&op); });
```
Comment (victor-eds) [inline polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp:454]:
```suggestion
        llvm::find_if(block,
                      [&](Operation &op) { return stores.contains(&op); });
```
NIT

**Review by whitneywhtsang — APPROVED** [review]

**Comment by sommerlukas** [comment]
Failures same as in https://github.com/intel/llvm/pull/10196, so unrelated to the changes here.

**Pull request merged**